pyriemann/classification.py

```diff
@@ -511,6 +516,36 @@ def predict(self, covtest):
             the prediction for each trials according to the closest centroid.
         """
         dist = self._predict_distances(covtest)
-        neighbors_classes = self.classes_[np.argsort(dist)]
+        neighbors_classes = self.classmeans_[np.argsort(dist)]
         out, _ = stats.mode(neighbors_classes[:, 0:self.n_neighbors], axis=1)
         return out.ravel()
+
+    def predict_proba(self, X):
+        """Predict proba using softmax.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_matrices, n_channels, n_channels)
+            Set of SPD matrices.
+
+        Returns
+        -------
+        prob : ndarray, shape (n_matrices, n_classes)
+            Probabilities for each class.
+        """
+        n_matrices, _, _ = X.shape
+
+        dist = self._predict_distances(X)
+        idx = np.argsort(dist)
+        dist_sorted = np.take_along_axis(dist, idx, axis=1)
+        neighbors_classes = self.classmeans_[idx]
+        probas = softmax(-dist_sorted[:, 0:self.n_neighbors]**2)
+
+        prob = np.zeros((n_matrices, len(self.classes_)))
+        for m in range(n_matrices):
+            for il, ll in enumerate(self.classes_):
+                prob[m, il] = np.sum(
+                    probas[m, neighbors_classes[m, 0:self.n_neighbors] == ll]
+                )
```

Comment:
Can you see how it differs from how sklearn does it?

https://github.com/scikit-learn/scikit-learn/blob/baf828ca126bcb2c0ad813226963621cafe38adb/sklearn/neighbors/_classification.py#L240

Reply:
Yes, I had checked: methods are very similar.

The difference is the definition of probabilities from distances:
pyRiemann uses a softmax of squared distances (this formula is derived from the Riemannian Gaussian distribution, as explained in #100),
whereas sklearn uses reciprocal of distances, and then divides by the sum (to have the sum of proba equal to one) (but, I don't know the origin of this formula).

Reply:
before committing and making a choice that differs from sklearn I would run a tiny benchmark on Gaussian data following LDA model to see what approach leads to the best calibrated probability.

Reply:
> I would run a tiny benchmark

Nice idea! But, who is "I"? you or me? ;-)

Reply:
Joking aside, I don't see how to do this benchmark.
Because inputs of sklearn are multivariate vectors (generated by a mixture of multivariate Gaussian distributions, ok), while inputs of pyRiemann are covariance matrices. So, results can't be compared.

Reply:
> I would do it in sklearn using euclidian data. Basically replace the
predict_proba in sklearn and see what works best
in sklearn. Then copy this in PyRiemann

I plot log-probabilities of kNN applied on bivariate Gaussian distributions:
- weights="distance" computing reciprocal of distances (classical sklearn implem);
- weights="max_lik", computing softmax of squared distances, a formula derived by maximum likelihood from a (Euclidean) Gaussian distribution (implem added in this branch).

Results are really close, but I think that new option is better, because derived from MLE.
Moreover, this option it could be added in sklearn.

![test_knn_comparison](https://user-images.githubusercontent.com/5346669/169088665-d4817e78-e341-44cd-be3d-01ed1e36cc7c.png)

Reply:
@agramfort , can we merge the branch?